From c240db5ed317fdc94532dc55d356f18306ded996 Mon Sep 17 00:00:00 2001
From: Pawel Raasz <pawel.raasz@intel.com>
Date: Tue, 8 Apr 2025 19:35:47 +0200
Subject: [PATCH] [build] Fix build issue using Fedora and gcc-15 (#29880)

### Details:
- Fix build issues for OpenVINO when using docker image with Fedora and
gcc-15

### Tickets:
 - CVS-164531
 - #29410

---------
Upstream-Status: Backport
[https://github.com/openvinotoolkit/openvino/commit/c240db5ed317fdc94532dc55d356f18306ded996]

Signed-off-by: Raasz, Pawel <pawel.raasz@intel.com>
---
 samples/cpp/common/format_reader/src/bmp.cpp         |  1 +
 src/common/util/src/wstring_convert_util.cpp         | 10 +++++-----
 src/core/dev_api/openvino/core/type/nf4.hpp          |  1 +
 src/core/include/openvino/core/type/bfloat16.hpp     |  1 +
 src/core/include/openvino/core/type/float16.hpp      |  1 +
 src/core/include/openvino/core/type/float4_e2m1.hpp  |  1 +
 src/core/include/openvino/core/type/float8_e4m3.hpp  |  1 +
 src/core/include/openvino/core/type/float8_e5m2.hpp  |  1 +
 src/core/include/openvino/core/type/float8_e8m0.hpp  |  1 +
 .../include/openvino/reference/gru_cell.hpp          | 12 +++++-------
 src/plugins/intel_cpu/src/nodes/rnn.cpp              |  3 ++-
 .../intel_cpu/src/shape_inference/custom/one_hot.cpp |  6 +++++-
 .../intel_gpu/src/graph/impls/cpu/proposal.cpp       |  1 +
 .../src/plugin/npuw/partitioning/online/graph.hpp    |  1 +
 .../intel_npu/src/plugin/npuw/serialization.hpp      |  1 +
 15 files changed, 28 insertions(+), 14 deletions(-)

diff --git a/samples/cpp/common/format_reader/src/bmp.cpp b/samples/cpp/common/format_reader/src/bmp.cpp
index b025e8992f..9d4cae1bc7 100644
--- a/samples/cpp/common/format_reader/src/bmp.cpp
+++ b/samples/cpp/common/format_reader/src/bmp.cpp
@@ -5,6 +5,7 @@
 // clang-format off
 #include <fstream>
 #include <iostream>
+#include <cstdint>
 
 #include "bmp.h"
 // clang-format on
diff --git a/src/common/util/src/wstring_convert_util.cpp b/src/common/util/src/wstring_convert_util.cpp
index de76e1134e..a841109276 100644
--- a/src/common/util/src/wstring_convert_util.cpp
+++ b/src/common/util/src/wstring_convert_util.cpp
@@ -13,9 +13,9 @@
 #        include <windows.h>
 #    endif
 
-#    ifdef __clang__
-#        pragma clang diagnostic push
-#        pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#    if defined(__clang__) || defined(__GNUC__)
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #    endif
 
 std::string ov::util::wstring_to_string(const std::wstring& wstr) {
@@ -45,8 +45,8 @@ std::wstring ov::util::string_to_wstring(const std::string& string) {
 #    endif
 }
 
-#    ifdef __clang__
-#        pragma clang diagnostic pop
+#    if defined(__clang__) || defined(__GNUC__)
+#        pragma GCC diagnostic pop
 #    endif
 
 #endif  // OPENVINO_ENABLE_UNICODE_PATH_SUPPORT
diff --git a/src/core/dev_api/openvino/core/type/nf4.hpp b/src/core/dev_api/openvino/core/type/nf4.hpp
index 383d3a01b2..685dffcc57 100644
--- a/src/core/dev_api/openvino/core/type/nf4.hpp
+++ b/src/core/dev_api/openvino/core/type/nf4.hpp
@@ -6,6 +6,7 @@
 
 #include <array>
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/bfloat16.hpp b/src/core/include/openvino/core/type/bfloat16.hpp
index 452239fb0d..be42942742 100644
--- a/src/core/include/openvino/core/type/bfloat16.hpp
+++ b/src/core/include/openvino/core/type/bfloat16.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/float16.hpp b/src/core/include/openvino/core/type/float16.hpp
index 31f7e421dd..05ef2b7fd4 100644
--- a/src/core/include/openvino/core/type/float16.hpp
+++ b/src/core/include/openvino/core/type/float16.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/float4_e2m1.hpp b/src/core/include/openvino/core/type/float4_e2m1.hpp
index b4e9f02992..f21b59df39 100644
--- a/src/core/include/openvino/core/type/float4_e2m1.hpp
+++ b/src/core/include/openvino/core/type/float4_e2m1.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/float8_e4m3.hpp b/src/core/include/openvino/core/type/float8_e4m3.hpp
index 166d0d15e8..ab209cf87f 100644
--- a/src/core/include/openvino/core/type/float8_e4m3.hpp
+++ b/src/core/include/openvino/core/type/float8_e4m3.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/float8_e5m2.hpp b/src/core/include/openvino/core/type/float8_e5m2.hpp
index e9e942bfb8..9c071b4d9d 100644
--- a/src/core/include/openvino/core/type/float8_e5m2.hpp
+++ b/src/core/include/openvino/core/type/float8_e5m2.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/include/openvino/core/type/float8_e8m0.hpp b/src/core/include/openvino/core/type/float8_e8m0.hpp
index ffc19134b3..7aabcac0c1 100644
--- a/src/core/include/openvino/core/type/float8_e8m0.hpp
+++ b/src/core/include/openvino/core/type/float8_e8m0.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <cmath>
+#include <cstdint>
 #include <iostream>
 #include <limits>
 #include <memory>
diff --git a/src/core/reference/include/openvino/reference/gru_cell.hpp b/src/core/reference/include/openvino/reference/gru_cell.hpp
index 4998d71cdc..b8129dbb52 100644
--- a/src/core/reference/include/openvino/reference/gru_cell.hpp
+++ b/src/core/reference/include/openvino/reference/gru_cell.hpp
@@ -105,14 +105,12 @@ void gru_cell(const T* X,
                                        reinterpret_cast<char*>(Ht_R_zrh[1].data()),
                                        reinterpret_cast<char*>(Ht_R_zrh[2].data())};
 
-    size_t num_b_splits = linear_before_reset ? 4 : 3;
+    const size_t num_b_splits = linear_before_reset ? 4 : 3;
     std::vector<std::vector<T>> biases_zrh(num_b_splits, std::vector<T>(B_shape[0] / num_b_splits));
-    std::vector<char*> pointers_biases = {reinterpret_cast<char*>(biases_zrh[0].data()),
-                                          reinterpret_cast<char*>(biases_zrh[1].data()),
-                                          reinterpret_cast<char*>(biases_zrh[2].data())};
-    if (linear_before_reset) {
-        pointers_biases.push_back(reinterpret_cast<char*>(biases_zrh[3].data()));
-    }
+    std::vector<char*> pointers_biases(num_b_splits);
+    std::transform(biases_zrh.begin(), biases_zrh.end(), pointers_biases.begin(), [](auto&& vec) {
+        return reinterpret_cast<char*>(vec.data());
+    });
 
     // split on gates
     reference::split(reinterpret_cast<char*>(Xt_W.data()), all_gates_shape, sizeof(T), 1, 3, pointers_XW.data());
diff --git a/src/plugins/intel_cpu/src/nodes/rnn.cpp b/src/plugins/intel_cpu/src/nodes/rnn.cpp
index 93d8b521a7..3c4a89af5a 100644
--- a/src/plugins/intel_cpu/src/nodes/rnn.cpp
+++ b/src/plugins/intel_cpu/src/nodes/rnn.cpp
@@ -280,7 +280,8 @@ bool RNN::isSupportedOperation(const std::shared_ptr<const ov::Node>& op, std::s
                               ov::op::v5::GRUSequence::get_type_info_static(),
                               ov::op::internal::AUGRUCell::get_type_info_static(),
                               ov::op::internal::AUGRUSequence::get_type_info_static())) {
-                if (rnnCellBase->get_activations() != std::vector<std::string>{"sigmoid", "tanh"}) {
+                if (const auto activations = std::vector<std::string>{"sigmoid", "tanh"};
+                    rnnCellBase->get_activations() != activations) {
                     errorMessage = "Not supported activation functions";
                     return false;
                 }
diff --git a/src/plugins/intel_cpu/src/shape_inference/custom/one_hot.cpp b/src/plugins/intel_cpu/src/shape_inference/custom/one_hot.cpp
index b97b2d47c3..0f9d73721f 100644
--- a/src/plugins/intel_cpu/src/shape_inference/custom/one_hot.cpp
+++ b/src/plugins/intel_cpu/src/shape_inference/custom/one_hot.cpp
@@ -21,7 +21,11 @@ Result OneHotShapeInfer::infer(const std::vector<std::reference_wrapper<const Ve
         OPENVINO_THROW("OneHot depth value can't be negative.");
     }
     auto result = input_shapes.front().get();
-    result.insert(result.begin() + m_axis, depth);
+    auto depth_pos = result.begin();
+    if (result.size() > 0) {
+        depth_pos += m_axis;
+    }
+    result.insert(depth_pos, depth);
 
     return {{std::move(result)}, ShapeInferStatus::success};
 }
diff --git a/src/plugins/intel_gpu/src/graph/impls/cpu/proposal.cpp b/src/plugins/intel_gpu/src/graph/impls/cpu/proposal.cpp
index c53fedba72..d86d6e8f9c 100644
--- a/src/plugins/intel_gpu/src/graph/impls/cpu/proposal.cpp
+++ b/src/plugins/intel_gpu/src/graph/impls/cpu/proposal.cpp
@@ -185,6 +185,7 @@ struct im_info_t {
 
 struct proposal_impl : typed_primitive_impl<proposal> {
     using parent = typed_primitive_impl<proposal>;
+    using parent::execute;
     using parent::parent;
 
     proposal_impl() : parent() {}
diff --git a/src/plugins/intel_npu/src/plugin/npuw/partitioning/online/graph.hpp b/src/plugins/intel_npu/src/plugin/npuw/partitioning/online/graph.hpp
index 3452a7e12a..4e47ee401e 100644
--- a/src/plugins/intel_npu/src/plugin/npuw/partitioning/online/graph.hpp
+++ b/src/plugins/intel_npu/src/plugin/npuw/partitioning/online/graph.hpp
@@ -6,6 +6,7 @@
 #pragma once
 
 #include <any>
+#include <cstdint>
 #include <functional>
 #include <iostream>
 #include <memory>
diff --git a/src/plugins/intel_npu/src/plugin/npuw/serialization.hpp b/src/plugins/intel_npu/src/plugin/npuw/serialization.hpp
index 88684dc33f..6cc1d2d6f9 100644
--- a/src/plugins/intel_npu/src/plugin/npuw/serialization.hpp
+++ b/src/plugins/intel_npu/src/plugin/npuw/serialization.hpp
@@ -5,6 +5,7 @@
 #pragma once
 
 #include <array>
+#include <cstdint>
 #include <functional>
 #include <iostream>
 #include <map>
-- 
2.43.0

